modules/validations/duplicate_address.js

```diff
@@ -0,0 +1,131 @@
+import { t } from '../util/locale';
+import { utilArrayIntersection, utilDisplayLabel } from '../util';
+import { validationIssue } from '../core/validation';
+
+
+export function validationDuplicateAddress(context) {
+  const type = 'duplicate_address';
+
+
+  function hasAddress(entity) {
+    return Object.keys(entity.tags).some(k => /^addr:/.test(k));
+  }
+
+
+  function isDuplicate(entity1, entity2) {
+    if (entity1.id === entity2.id) return false;   // ignore self
+
+    let dupe = { entity1: entity1, entity2: entity2, tags: {} };
+    let different = false;
+
+    // if any of these tags don't match, *and both are present*, it's not a duplicate..
+    ['addr:street', 'addr:place', 'addr:block'].forEach(k => {
+      const val1 = entity1.tags[k] || '';
+      const val2 = entity2.tags[k] || '';
+      if (val1 && val2) {     // both assigned
+        if (val1 !== val2) {
+          different = true;   // values don't match
+        } else {
+          dupe.tags[k] = val1;
+        }
+      }
```

Comment:
what's the idea here behind the case of "only one of them is present"? do we still consider them as "potentially dupe" because they are so close to each other?

Reply:
We definitely want to avoid user confusion, but the `reference` message appears above a table of actual duplicated tags that the code caught (see the screenshot) - maybe does this provide the context well enough?

For the "only one present" vs "both present" comments in the code, it is because none of the tags are really guaranteed to exist.  
* If one address has a `addr:street` and an adjacent one doesn't have it, the code will still warn about duplicate housenumbers... 
* But if one address has `addr:unit` and an adjacent one doesn't, we would not issue that warning (the one without the unit is the building / front door and the one with the unit is interior address, this is ok).

Reply:
Sounds good! I was just asking about "our way of thinking" as engineers, not about how to explain different cases to the user in the "reference" message.
Also, since "none of them are guaranteed to exist", I think saying "the same house number" is enough for the reference message.